### DISCUSSION TITLE
How to enable Docker layer caching between builds in GitHub actions?

### DISCUSSION THREAD

**Discussion by sschuberth** [batect/batect#1245] [Q&A]
I was trying to [speed up our Batect runs across PRs](https://github.com/oss-review-toolkit/ort/pull/5070), where we're using Batect to setup an environment for functional tests. If nothing in the test environment / Dockerfile changes, my goal would be that Batect's setup across PRs would essentially be a no-op as everything comes from the layer cache.

However, guessing from the [build time for `funTest-analyzer-docker` of a whopping 41 minutes](https://github.com/oss-review-toolkit/ort/actions/runs/1859548273), which is about twice as much as expected, Batect does not seem to make use of the layers restored via `docker/build-push-action`, and instead the Docker image was built twice.

So I'm wondering whether anyone was able to setup layer changing for Batect via https://github.com/docker/build-push-action/blob/master/docs/advanced/cache.md#github-cache?

**Comment by charleskorn** [comment]
Do you see the same behaviour (slower than expected image builds) when using `docker build ...` to build your image?

Also, if you're wanting to use the [cache backend](https://github.com/docker/build-push-action/blob/master/docs/advanced/cache.md#cache-backend-api), keep in mind that you need to use `cache-from` and `cache-to`, which Batect doesn't yet support.

**Reply by sschuberth** [reply]
> Also, if you're wanting to use the [cache backend](https://github.com/docker/build-push-action/blob/master/docs/advanced/cache.md#cache-backend-api), keep in mind that you need to use `cache-from` and `cache-to`

Hmm... I'm not sure whether I get what you're trying to say. My understanding is that if I use

    - name: Build the Docker image
      uses: docker/build-push-action@v2
      with:
        context: .
        cache-from: type=gha
        cache-to: type=gha,mode=max

then the `build-push-action` takes care of restoring the Docker layers in a way that is transparent for all "regular" Docker uses, incl. running Batect. So I would expect Batect to benefit from the restored layers. Is this assumption of mine wrong?

**Reply by charleskorn** [reply]
I'm not sure about that - if you run `docker build .` immediately after that step, does it reuse the restored layers? If it does reuse the layers and Batect does not reuse the layers, then that sounds like an issue in Batect.

**Reply by charleskorn** [reply]
Another thing to try would be to enable more verbose output from Batect (`--output all`) and see if the output from the image build shows it rebuilding the image or reusing the existing layers.

**Reply by sschuberth** [reply]
> if you run `docker build .` immediately after that step, does it reuse the restored layers?

It seems so, yes. See how [here](https://github.com/oss-review-toolkit/ort/runs/8056325185?check_suite_focus=true#step:5:3319) the "Build the Docker image" step takes 10 minutes whereas [here](https://github.com/oss-review-toolkit/ort/runs/8056553739?check_suite_focus=true#step:5:199) in a subsequent run it takes only 8 seconds thanks to the reuse of cached layers.

*Edit:* I just realized that comparison to be incorrect, as it exclusively uses the `docker/build-push-action` which naturally *can* restore `cache-from: type=gha`. I'll do another test with running `docker build .` manually.

*Edit 2:* Indeed, this run shows that while a subsequent run of `docker/build-push-action` [does get sped up](https://github.com/oss-review-toolkit/ort/runs/8065322397?check_suite_focus=true#step:5:190) thanks to caching, the manual `docker build .` run [does not](https://github.com/oss-review-toolkit/ort/runs/8065322397?check_suite_focus=true#step:6:3211):

![image](https://user-images.githubusercontent.com/349154/187148477-b3d5cd42-7e24-4945-9247-e3f765b4766f.png)

**Reply by sschuberth** [reply]
I've asked about guidance at https://github.com/docker/build-push-action/issues/681.

**Reply by charleskorn** [reply]
OK, that's the behaviour I expected. Adding support for cache-from and cache-to isn't particularly difficult now - watch this space :)

**Reply by sschuberth** [reply]
For completeness, I've filed https://github.com/batect/docker-client/issues/244 about this.

**Comment by colt-jay** [comment]
> you need to use cache-from and cache-to, which Batect doesn't yet support.

@charleskorn I've reviewed the roadmap, issues, and PRs and it looks like this is not a priority right now right? 

To make sure I understand, you suggest that we pre-build the images using `docker build` + the buildx caching mechanism so that batect can then build from that image?

**Reply by charleskorn** [reply]
Support for cache-from and cache-to will be added as soon as I have the new Docker client in place, which is my current focus area. I can't commit to a particular timeframe, but it will be soon :)

> To make sure I understand, you suggest that we pre-build the images using docker build + the buildx caching mechanism so that batect can then build from that image?

That _should_ work - Docker should pull down the layers it can reuse from the cache while it's building the image, and then Batect should just reuse the already-built image. If this isn't working for you, try running Batect with `--output=all` to get the output from the image build process to see if it's reusing layers from the local cache or not.

**Reply by sschuberth** [reply]
> Support for cache-from and cache-to will be added as soon as I have the new Docker client in place

So, should that be the case with Batect 0.80.0 now?

**Reply by charleskorn** [reply]
Not yet :) But now that the new Docker client is in place, adding support for cache-from and cache-to should be straightforward.

**Reply by sschuberth** [reply]
I'm still wondering why there needs to be explicit support for `cache-from` / `cache-to` in Batect's Docker client. My understanding is [still](https://github.com/batect/batect/discussions/1245#discussioncomment-2637591) that the `docker/build-push-action` *already* restores Docker layers from the GitHub Action cache, and any tool that uses Docker afterwards should benefit from those cached layers...

*Edit:* I just realized that I never fully replied to your follow-up [question above](https://github.com/batect/batect/discussions/1245#discussioncomment-2650491). Will do that now.

**Comment by sschuberth** [comment]
Meanwhile, I was also trying to pass `load: true` as suggested [here](https://docs.docker.com/build/ci/github-actions/export-docker/) to `docker/build-push-action` as the docs sounded as if that would make restored layers cached in `gha` available to CLI commands like `docker build` (and thus hopefully also to `batect`), but I couldn't get it to work. Images / layers seems to be rebuilt all the time.
